### ISSUE TITLE
RFC: mounting MachineLabs Datasets

### ISSUE THREAD

**Issue by cburgdorf** [machinelabs/machinelabs#564]
### Summary

This issue describes a new mechanism for users to get hold of datasets.

### Motivation

As a user I want to work with datasets to train a model.

So far MachineLabs offers two ways to get a hold of datasets:

1. Download data from the internet
2. Put data write into source files

While the second has very narrow use cases (mainly, generating data on the fly), the first one already opens a fair range of use cases to do some real training.

However, there are a couple of important problems with the first approach.

1. Downloading data from the internet takes time. With our current model, the user has to re-download the data with every execution. If the dataset is 10 GB it's unpractical for the user to re-download the dataset for every little code change that the user does (e.g. fixing syntax erros).

2. We have to limit the amount of space that a user can download or even just consumes in other ways. Therefore, each execution container is already limited to 5 gig.

3. We have to minimize the bandwith we consume in general. E.g. every user downloading MNIST over and over is bad.

4. We need a way for anonymous users to consume data without giving such users internet access. This is a requirement to get into the keras docs.

#### Introducing MachineLabs Datasets

MachineLabs Datasets will give each user an online storage to put datasets up which can be mounted at execution time. The data can be made *public* so that other users can also consume it or *private* so that just the owner can see and consume it.

#### How to upload the data.

The user should be able to upload data through the MachineLabs UI and also manage the files from there. 

I also believe that it may be even more important to give the user the option to tell us where to download the file from rather than to actual upload it from the users machine.

For instance, if I want to put the dataset for the fast style transfer into my drive, I'd rather like to be able to tell MachineLabs to take it from `http://msvocds.blob.core.windows.net/coco2014/train2014.zip` instead of forcing the user to first download it to their local machine, just to upload it from the local machine to our servers.

So, in my view, we actually even want to start with this alternative "download from URL" option.

#### How should data be organized?

Notice that we don't just open up some space for the user to write (think GDrive, Dropbox) but instead give them a well structured way to access datasets. 

1. There's a well recognizable pattern of `/username/dataset/version` so it's not that one user has data at `/user-foo/my/stuff/goes/here/mnist-v1` and another one has `/user-bar/mnist/1`. Instead, everyone will follow the schema `/username/datasets/mnist/1`. 

2. With this schema, managing visibility (public / private) becomes simple because we can keep each dataset as a well formed entity with metadata in our firebase. Managing privacy would be very tricky if each user has absolute freedom about how to organize their online space. 

#### How to mount the data into an execution

Mounting the data into an execution can be done via `ml.yaml` and should look familar to `inputs`

```
mounts:
   - path: /machinelabs/datasets/mnist
     name: mnist
  - path: /some-user/datasets/whatever
    name: whatever
```

The above configuration would result in two mounted directories with one being at `/ml-drive/mnist` and one at `/ml-drive/whatever`. 

Notice that all mounts should always get mounted at `/ml-drive` to reduce the possibility of failures. (e.g. the user can't try to mount something at `/mnt`, `/etc` or `/dev` etc).

#### Mount validation

As mentioned before, we should make sure that we are prepared for private datasets. That is, datasets that only the owner can see and use. One important thing to get right here is that the path should not change based on the fact if some user directory in the drive is marked *public* or *private*.

In other words if `/user-foo/dataset/something` was *private* so far, and now the user decided to make it public, it would still be at `/user-foo/dataset/something`.

We can achieve that if we always validate the mounted `path` before we actually mount it into the container. In other words, we never blindly mount something into the bucket without validating the specific metadata first.

Notice also that the `path` must be a valid dataset path. That is, it is not allowed to directly mount `/user-foo/datasets` for instance.
